openmc/dagmc.py

```diff
@@ -0,0 +1,593 @@
+        id = int(group.name.split('/')[-1].lstrip('universe '))
+        fname = group['filename'][()].decode()
+        name = group['name'][()].decode() if 'name' in group else None
+
+        out = cls(fname, universe_id=id, name=name)
+
+        out.auto_geom_ids = bool(group.attrs['auto_geom_ids'])
+        out.auto_mat_ids = bool(group.attrs['auto_mat_ids'])
+
+        return out
+
+    @classmethod
+    def from_xml_element(cls, elem):
+        """Generate DAGMC universe from XML element
+
+        Parameters
+        ----------
+        elem : lxml.etree._Element
+            `<dagmc_universe>` element
+
+        Returns
+        -------
+        openmc.DAGMCUniverse
+            DAGMCUniverse instance
+
+        """
+        id = int(get_text(elem, 'id'))
+        fname = get_text(elem, 'filename')
+
+        out = cls(fname, universe_id=id)
+
+        name = get_text(elem, 'name')
+        if name is not None:
+            out.name = name
+
+        out.auto_geom_ids = bool(elem.get('auto_geom_ids'))
+        out.auto_mat_ids = bool(elem.get('auto_mat_ids'))
+
+        for item in elem.find('material_overrides').attrib:
+            origin_mat, overwrite = item
+            for mat_name in overwrite.split():
+                out.material_overrides.setdefault(
+                            origin_mat.lower(), []).append(mat_name)
+
+        return out
+
+    def _partial_deepcopy(self):
+        """Clone all of the openmc.DAGMCUniverse object's attributes except for
+        its cells, as they are copied within the clone function. This should
+        only to be used within the openmc.UniverseBase.clone() context.
+        """
+        clone = openmc.DAGMCUniverse(name=self.name, filename=self.filename)
+        clone.volume = self.volume
+        clone.auto_geom_ids = self.auto_geom_ids
+        clone.auto_mat_ids = self.auto_mat_ids
+        return clone
+
+    def add_cell(self, cell):
+        """Add a cell to the universe.
+
+        Parameters
+        ----------
+        cell : openmc.DAGMCCell
+            Cell to add
+
+        """
+        if not isinstance(cell, openmc.DAGMCCell):
+            msg = f'Unable to add a DAGMCCell to DAGMCUniverse '  \
+                  f'ID="{self._id}" since "{cell}" is not a DAGMCCell'
+            raise TypeError(msg)
+
+        cell_id = cell.id
+
+        if cell_id not in self._cells:
+            self._cells[cell_id] = cell
+
+    def remove_cell(self, cell):
+        """Remove a cell from the universe.
+
+        Parameters
+        ----------
+        cell : openmc.Cell
+            Cell to remove
+
+        """
+
+        if not isinstance(cell, openmc.DAGMCCell):
+            msg = f'Unable to remove a Cell from Universe ID="{self._id}" ' \
+                f'since "{cell}" is not a Cell'
+            raise TypeError(msg)
+
+        # If the Cell is in the Universe's list of Cells, delete it
+        self._cells.pop(cell.id, None)
+
+    def sync_dagmc_cells(self, mats={}):
```

Comment:
Should the `mats` parameter really be optional? It seems this method would fail without a valid entry.